### PR TITLE
Upgrade SB3 and migrate to local logger

### LINKED ISSUE #325: Upgrade stable-baselines3
We are currently on `stable-baselines3~=0.10.0` while the newest version is `1.1.0`. There have been a lot of bug fixes since then (see https://github.com/DLR-RM/stable-baselines3/releases) and in particular, [issue #109](https://github.com/DLR-RM/stable-baselines3/issues/109) has been resolved, which would remove hacks in a number of places in `imitation` (just search for `#109` in the `imitation` code).

There are some breaking changes from `0.10.0` to `1.1.0`, we at least have to slightly modify logging in a few places because the global logger has been removed. Not sure if any of the other changes affect `imitation`.

### PR DESCRIPTION
**Pull request by AdamGleave**
Move from SB3 0.10.0 to latest version, currently 1.1.0.

SB3 moved from a global logger instance to a local logger for each algorithm. This PR switches to the same style.

I've also tidied up some hacks that we had to make things work with the global logger previously. I'm not sure I've caught all of them though. In particular there were some cases where `verbose=1` was flagged with a FIXME as being a workaround, but `verbose=1` also appears elsewhere in the codebase. I'm not sure whether these were introduced as a workaround or because we actually want that verbosity.

Overall I think this PR should be LOC negative :)

@qxcv: one tricky thing that came up here was that DAgger is pickled to save it (as is, indirectly, BC), but the Logger is not pickleable. I've just removed the Logger from pickle, and setup a default logger on restore. This feels a bit clunky though. I think you've got a better sense than me when DAgger needs to be saved/restored, so if you have a better idea for this let me know.

Fixes #325 and #327.

### CODE REVIEW

**Comment by qxcv** [comment]
> @qxcv: one tricky thing that came up here was that DAgger is pickled to save it (as is, indirectly, BC), but the Logger is not pickleable. I've just removed the Logger from pickle, and setup a default logger on restore. This feels a bit clunky though. I think you've got a better sense than me when DAgger needs to be saved/restored, so if you have a better idea for this let me know.

That seems fine to me! I guess the only downside is that logs will be stored separately for each restore/save cycle?

**Comment by AdamGleave** [comment]
> That seems fine to me! I guess the only downside is that logs will be stored separately for each restore/save cycle?

Yeah, unless the user re-configures the logger to have the same directory at least. (And even then I'm not sure things are safe to re-open?)

We could mess around to make `Logger` pickleable, in terms of e.g. saving the folder and any other state needed to restore (tensorboard step?), but restoration would then break across hosts/filesystems which seems bad in its own way.

**Review by ejnnr — APPROVED** [review]
LGTM except for the small things I commented on (which I don't think require re-review after addressing them).

Two things that probably don't need to be addressed in this PR:
- I agree refactoring to get rid of `utils.init_rl` sounds good (though I haven't checked in detail whether there would be any significant hurdles). Apart from making the code simpler to read, having a CLI argument like `rl_kwargs` or `agent_kwargs` is probably more intuitive than the current `init_rl_kwargs`
- My type checker (Pylance) complains about a couple of type issues and I think it's right about some of them. For example, we should probably use `output_formats=[]` here? https://github.com/HumanCompatibleAI/imitation/blob/dd1b747714fc42d44261931fc88b1022ce23d296/src/imitation/util/logger.py#L51
  And there are a couple of places where we should technically convert an `AnyPath` to `str` before passing it on. But I think all of them already existed before the PR anyway.

File: src/imitation/util/logger.py [hunk 1]
```diff
@@ -160,55 +144,15 @@ def configure(
...
-    # Replace `stable_baselines3.logger` methods with erroring stubs to
-    # prevent unexpected logging state from mixed logging configuration.
-    sb_logger.configure = _sb_logger_configure_replacement
-    sb_logger.reset = _sb_logger_reset_replacement
-
+    if folder is None:
+        if folder is None:
+            now = datetime.datetime.now()
+            timestamp = now.strftime("imitation-%Y-%m-%d-%H-%M-%S-%f")
+            folder = os.path.join(tempfile.gettempdir(), timestamp)
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/util/logger.py:151]: Doh, fixed.

File: src/imitation/algorithms/bc.py [hunk 2]
```diff
@@ -414,3 +418,14 @@ def save_policy(self, policy_path: types.AnyPath) -> None:
...
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        # logger can't be pickled as it depends on open files
+        del state["logger"]
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        # callee should call set_logger if they want to override this
+        self._logger = state.get("logger") or logger.configure()
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/algorithms/bc.py:431]: oops, yes

File: setup.py [hunk 3]
```diff
@@ -56,7 +56,7 @@ def get_readme() -> str:
         "torch>=1.4.0",
         "tqdm",
         "scikit-learn>=0.21.2",
-        "stable-baselines3~=0.10.0",
+        "stable-baselines3",
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply setup.py:59]: Good point, changed.

File: src/imitation/algorithms/dagger.py [hunk 4]
```diff
@@ -314,12 +315,29 @@ def __init__(
         self._last_loaded_round = -1
         self._all_demos = []
 
+        self._logger = custom_logger or logger.configure()
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/algorithms/dagger.py:318]: It is deliberate. We need to set the BC logger from DAgger so I actually want `logger` exposed for BC. Granted the inconsistency is grating, so maybe we should go with `_logger` and `set_logger` for both, but on balance I think additional complexity isn't worth it.

**Comment by AdamGleave** [comment]
Thanks for the review!

I've opened https://github.com/HumanCompatibleAI/imitation/issues/327 to track `init_rl` removal.

Agree there are some type errors. I've introduced a new function to convert `AnyPath` to `str`.

Also added some missing functions in `Logger` and expanded test coverage.

**Pull request merged**